feat: handle gmp message for leverage messages

gsk967 opened the pull request:
<!-- markdownlint-disable MD041 -->

## Description

closes: #XXXX

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **Refactor**
	- Enhanced memo handling in packet reception for improved reliability and error logging.
	- Streamlined memo processing logic for better efficiency and maintainability.
	- Improved message type cases for better performance.
	- Removed unnecessary functions to simplify code structure.
	- Updated memo handling logic in `onRecvPacketPrepare` method for improved processing.
	- Adjusted handling of `memo` and `memoPayload` for better deserialization.
	- Commented out GMP handling section in the `execute` method for potential processing changes.
	- Renamed `Message` struct to `Memo` for consistency in ICS20 packet memo field.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

robert-zaremba reviewed (commented):
1. gmp handling must be done after the transfer - otherwise, if we execute a bank `Send`, before the transfer is recorded, we will miss the funds
2. Need to add checks as we do in the normal hooks
3. If we handle GMP, then we must abort the "normal" hooks execution
4. Need to add GMP types and support fallback.

gsk967 commented on x/uibc/uics20/memo_handler.go:
```diff
@@ -105,10 +112,10 @@ func (mh MemoHandler) execute(ctx *sdk.Context) error {
-		gh := gmp.NewHandler()
-		return gh.OnRecvPacket(*ctx, mh.received, mh.memo, mh.receiver)
-	}
+	// if mh.isGMP {
```
No need for separate execution for GMP handler because we already preprocessing the gmp handler memo on `mh.onRecvPacketPrepare`

gsk967 commented on x/uibc/uics20/memo_handler.go:
```diff
@@ -58,13 +58,20 @@ func (mh *MemoHandler) onRecvPacketPrepare(
 		events = append(events, "axelar GMP transaction")
 		mh.isGMP = true
-		return events, nil
+		gh := gmp.NewHandler()
```
Now gmp incoming payload will be assigned to memoPayload because it is same as normal ibc memo , so we are not doing any separate execution for gmp handler it is same as ibc memo handler

gsk967 commented:
> 1. gmp handling must be done after the transfer - otherwise, if we execute a bank `Send`, before the transfer is recorded, we will miss the funds
> 
>     2. Need to add checks as we do in the normal hooks
> 
>     3. If we handle GMP, then we must abort the "normal" hooks execution
> 
>     4. Need to add GMP types and support fallback.

1. There is no need for separate execution for gmp handler because the incoming gmp payload is same as ibc memo and it is executing after normal ibc transfer execution 
2. The incoming payload same as ibc memo payload , so we already added checks for msgs 
3. No need to abort the hooks execution 
4. We already added GMP types https://github.com/umee-network/umee/blob/sai/handle_gmp_messages/x/uibc/gmp/types.go#L9-L15

robert-zaremba commented on x/uibc/uics20/memo_handler.go:
```diff
@@ -58,13 +58,20 @@ func (mh *MemoHandler) onRecvPacketPrepare(
 
+	var memoPayload []byte
 	if strings.EqualFold(ftData.Sender, gmp.DefaultGMPAddress) {
 		events = append(events, "axelar GMP transaction")
```
```suggestion
		events = append(events, "Axelar GMP")
```

robert-zaremba commented on x/uibc/uics20/memo_handler.go:
```diff
@@ -58,13 +58,20 @@ func (mh *MemoHandler) onRecvPacketPrepare(
 		events = append(events, "axelar GMP transaction")
 		mh.isGMP = true
-		return events, nil
+		gh := gmp.NewHandler()
+		gmpMessage, err := gh.OnRecvPacket(*ctx, mh.received, mh.memo, mh.receiver)
+		if err != nil {
```
```suggestion
		gmpMessage, err := gmp.ParseMemo(*ctx, mh.received, mh.memo, mh.receiver)
		if err != nil {
```

We don't need a handler struct. Let's simplify

robert-zaremba commented on x/uibc/uics20/memo_handler.go:
```diff
@@ -105,10 +112,10 @@ func (mh MemoHandler) execute(ctx *sdk.Context) error {
+	// if mh.isGMP {
+	// 	gh := gmp.NewHandler(mh.cdc)
+	// 	return gh.OnRecvPacket(ctx, mh.received, mh.memo, mh.receiver)
+	// }
```
let's remove this commented code

robert-zaremba commented on x/uibc/uics20/memo_handler.go:
```diff
@@ -58,13 +58,20 @@ func (mh *MemoHandler) onRecvPacketPrepare(
+		gh := gmp.NewHandler()
+		gmpMessage, err := gh.OnRecvPacket(*ctx, mh.received, mh.memo, mh.receiver)
+		if err != nil {
+			return events, err
```
1. let's log an error here
2. unmarshalling error nor validation error should not abort the transfer. Please update a code and add a test for that.

robert-zaremba reviewed (commented):
1. Need to fix error handling
2. Please add more tests, or list all missing tests in the subtask of the epic.

gsk967 commented on x/uibc/gmp/gmp_middleware_test.go:
```diff
@@ -40,13 +40,28 @@ func TestGmpMemoHandler(t *testing.T) {
+				validMemo := Message{
+					SourceChain:   "source_chain",
+					SourceAddress: "source_addr",
+					Payload:       nil,
```
Yes, it is payload is `[]byte` https://github.com/axelarnetwork/evm-cosmos-gmp-sample/blob/main/native-integration/sample-middleware/types.go#L18-L23

robert-zaremba commented on x/uibc/gmp/types.go:
```diff
@@ -15,8 +15,8 @@ const (
-// Message is attached in ICS20 packet memo field
-type Message struct {
+// GMPMemo is attached in ICS20 packet memo field
+type GMPMemo struct {
```
```suggestion
type Memo struct {
```
GMP is already a package name

robert-zaremba reviewed (approved):
Pre-approving
